kafka/consumer_test.go

```diff
@@ -18,12 +18,305 @@
+	assert.NoError(t, err)
+	assert.NotNil(t, r)
+
+	consumer := newConsumer(t, cfg)
+	assert.NoError(t, consumer.fetch(context.Background()))
+}
+
+func TestMultipleConsumers(t *testing.T) {
+	event := model.APMEvent{Transaction: &model.Transaction{ID: "1"}}
+	codec := json.JSON{}
+	topics := []string{"topic"}
+	client, addrs := newClusterWithTopics(t, topics...)
+
+	var count atomic.Int32
+	cfg := ConsumerConfig{
+		Brokers: addrs,
+		Topics:  topics,
+		GroupID: "groupid",
+		Decoder: codec,
+		Logger:  zap.NewNop(),
+		Processor: model.ProcessBatchFunc(func(ctx context.Context, b *model.Batch) error {
+			count.Add(1)
+			assert.Len(t, *b, 1)
+			//assert.Equal(t, event, (*b)[0])
+			return nil
+		}),
+	}
+
+	b, err := codec.Encode(event)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	consumer1 := newConsumer(t, cfg)
+	go func() {
+		assert.ErrorIs(t, consumer1.Run(ctx), context.Canceled)
+	}()
+
+	waitCh := make(chan struct{})
+	go func() {
+		defer close(waitCh)
+		for i := 0; i < 1000; i++ {
+			results := client.ProduceSync(context.Background(), &kgo.Record{
+				Topic: topics[0],
+				Value: b,
+			})
+			assert.NoError(t, results.FirstErr())
+			record, err := results.First()
+			assert.NoError(t, err)
+			assert.NotNil(t, record)
+		}
+	}()
+
+	consumer2 := newConsumer(t, cfg)
+	go func() {
+		assert.ErrorIs(t, consumer2.Run(ctx), context.Canceled)
+	}()
+
+	<-waitCh
+
+	assert.Eventually(t, func() bool {
+		return count.Load() == 1000
+	}, 1*time.Second, 50*time.Millisecond)
+}
+
+func TestMultipleConsumerGroups(t *testing.T) {
+	event := model.APMEvent{Transaction: &model.Transaction{ID: "1"}}
+	codec := json.JSON{}
+	topics := []string{"topic"}
+	client, addrs := newClusterWithTopics(t, topics...)
+
+	cfg := ConsumerConfig{
+		Brokers: addrs,
+		Topics:  topics,
+		Decoder: codec,
+		Logger:  zap.NewNop(),
+	}
+
+	b, err := codec.Encode(event)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	m := make(map[string]*atomic.Int64)
+
+	amount := 100
+
+	for i := 0; i < amount; i++ {
+		gid := "groupid" + strconv.Itoa(i)
+		cfg.GroupID = gid
+		m[gid] = &atomic.Int64{}
+		cfg.Processor = model.ProcessBatchFunc(func(ctx context.Context, b *model.Batch) error {
+			count := m[gid]
+			count.Add(1)
+			assert.Len(t, *b, 1)
+			assert.Equal(t, event, (*b)[0])
+			return nil
+		})
+		consumer := newConsumer(t, cfg)
+		go func() {
+			assert.ErrorIs(t, consumer.Run(ctx), context.Canceled)
+		}()
+	}
+
+	waitCh := make(chan struct{})
+	go func() {
+		defer close(waitCh)
+		for i := 0; i < amount; i++ {
+			client.Produce(context.Background(), &kgo.Record{
+				Topic: topics[0],
+				Value: b,
+			}, func(r *kgo.Record, err error) {
+				assert.NoError(t, err)
+				assert.NotNil(t, r)
+			})
+
+		}
+	}()
+
+	<-waitCh
+
+	for k, i := range m {
+		assert.Eventually(t, func() bool {
+			return i.Load() == int64(amount)
```

Comment:
It's not necessary to unit test 100 different consumer groups and tie the number of consumer groups to the number of produced records. I think it would be better to have two variables, one for the records produced and another for the number of consumer groups. Also there's no need to launch a goroutine for the producer and sync with a channel:

```suggestion
	groups := 10
	for i := 0; i < groups; i++ {
		gid := "groupid" + strconv.Itoa(i)
		cfg.GroupID = gid
		m[gid] = &atomic.Int64{}
		cfg.Processor = model.ProcessBatchFunc(func(ctx context.Context, b *model.Batch) error {
			count := m[gid]
			count.Add(1)
			assert.Len(t, *b, 1)
			assert.Equal(t, event, (*b)[0])
			return nil
		})
		consumer := newConsumer(t, cfg)
		go func() {
			assert.ErrorIs(t, consumer.Run(ctx), context.Canceled)
		}()
	}

	produceRecords := 100
	for i := 0; i < produceRecords; i++ {
		client.Produce(context.Background(), &kgo.Record{
			Topic: topics[0],
			Value: b,
		}, func(r *kgo.Record, err error) {
			assert.NoError(t, err)
			assert.NotNil(t, r)
		})
	}

	for k, i := range m {
		assert.Eventually(t, func() bool {
			return i.Load() == int64(produceRecords)
```